// docpad/out/lib/collections/files.js
// Generated by CoffeeScript 1.3.3
var FileModel, FilesCollection, Model, QueryCollection, balUtil, _, _ref,
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

_ = require('underscore');

balUtil = require('bal-util');

_ref = require(__dirname + '/../base'), QueryCollection = _ref.QueryCollection, Model = _ref.Model;

FileModel = require(__dirname + '/../models/file');

FilesCollection = (function(_super) {

  __extends(FilesCollection, _super);

  function FilesCollection() {
    return FilesCollection.__super__.constructor.apply(this, arguments);
  }

  FilesCollection.prototype.model = FileModel;

  FilesCollection.prototype.collection = FilesCollection;

  FilesCollection.prototype.comparator = [
    {
      date: -1,
      name: 1
    }
  ];

  return FilesCollection;

})(QueryCollection);

module.exports = FilesCollection;
